overlord/snapstate/handlers.go

```diff
@@ -3532,8 +3637,24 @@ func (m *SnapManager) doSetupAliases(t *state.Task, _ *tomb.Tomb) error {
 	}
 	snapName := snapsup.InstanceName()
 	curAliases := snapst.Aliases
+	autoDisabled := snapst.AutoAliasesDisabled
+
+	var oldAliases map[string]*AliasTarget
+	oldAutoDisabled := autoDis
 
-	_, _, err = applyAliasesChange(snapName, autoDis, nil, snapst.AutoAliasesDisabled, curAliases, m.backend, doApply)
+	prune, err := shouldPruneOldAliases(t, snapName)
```

Comment:
It would fail unfortunately, because of how `applyAliasesChange` is implemented, as it gets `prevAliases` from a parameter rather than from what is actually on disk for efficiency, so having `prevAlias == newAliases` is basically a no-op if the aliases are unchanged which is a problem if `remove-aliases` ran earlier because that wouldn't match what is actually on disk. This is why I try detect whether `remove-aliases` ran or not.

An alternative approach would be to scan for snap aliases on disk and add `Ensure*` like behavior maybe inside existing `backend.UpdateAliases` or create a separate `backend.EnsureAliases`, the disadvantage is that it would be inefficient because it would read all files under `SnapBinariesDir`  and do a `readlink` on each file.